// @tensorflow/tfjs-data/dist/iterators/lazy_iterator.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var tf = require("@tensorflow/tfjs-core");
var tensor_util_1 = require("@tensorflow/tfjs-core/dist/tensor_util");
var seedrandom = require("seedrandom");
var deep_map_1 = require("../util/deep_map");
var growing_ring_buffer_1 = require("../util/growing_ring_buffer");
var ring_buffer_1 = require("../util/ring_buffer");
function iteratorFromItems(items) {
    return new ArrayIterator(items);
}
exports.iteratorFromItems = iteratorFromItems;
function iteratorFromIncrementing(start) {
    var i = start;
    return iteratorFromFunction(function () { return ({ value: i++, done: false }); });
}
exports.iteratorFromIncrementing = iteratorFromIncrementing;
function iteratorFromFunction(func) {
    return new FunctionCallIterator(func);
}
exports.iteratorFromFunction = iteratorFromFunction;
function iteratorFromConcatenated(baseIterators, baseErrorHandler) {
    return new ChainedIterator(baseIterators, baseErrorHandler);
}
exports.iteratorFromConcatenated = iteratorFromConcatenated;
function iteratorFromConcatenatedFunction(iteratorFunc, count, baseErrorHandler) {
    return iteratorFromConcatenated(iteratorFromFunction(iteratorFunc).take(count), baseErrorHandler);
}
exports.iteratorFromConcatenatedFunction = iteratorFromConcatenatedFunction;
function iteratorFromZipped(iterators, mismatchMode) {
    if (mismatchMode === void 0) { mismatchMode = ZipMismatchMode.FAIL; }
    return new ZipIterator(iterators, mismatchMode);
}
exports.iteratorFromZipped = iteratorFromZipped;
var LazyIterator = (function () {
    function LazyIterator() {
    }
    LazyIterator.prototype.collect = function (maxItems, prefetch) {
        if (maxItems === void 0) { maxItems = 1000; }
        if (prefetch === void 0) { prefetch = 100; }
        return __awaiter(this, void 0, void 0, function () {
            var stream, result, count, x;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        stream = prefetch > 0 ? this.prefetch(prefetch) : this;
                        result = [];
                        count = 0;
                        return [4, stream.next()];
                    case 1:
                        x = _a.sent();
                        _a.label = 2;
                    case 2:
                        if (!!x.done) return [3, 4];
                        result.push(x.value);
                        count++;
                        if (count >= maxItems) {
                            return [2, result];
                        }
                        return [4, stream.next()];
                    case 3:
                        x = _a.sent();
                        return [3, 2];
                    case 4: return [2, result];
                }
            });
        });
    };
    LazyIterator.prototype.resolveFully = function () {
        return __awaiter(this, void 0, void 0, function () {
            var x;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.next()];
                    case 1:
                        x = _a.sent();
                        _a.label = 2;
                    case 2:
                        if (!!x.done) return [3, 4];
                        return [4, this.next()];
                    case 3:
                        x = _a.sent();
                        return [3, 2];
                    case 4: return [2];
                }
            });
        });
    };
    LazyIterator.prototype.resolveWhile = function (predicate) {
        return __awaiter(this, void 0, void 0, function () {
            var x, shouldContinue;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.next()];
                    case 1:
                        x = _a.sent();
                        shouldContinue = predicate(x.value);
                        _a.label = 2;
                    case 2:
                        if (!((!x.done) && shouldContinue)) return [3, 4];
                        return [4, this.next()];
                    case 3:
                        x = _a.sent();
                        shouldContinue = predicate(x.value);
                        return [3, 2];
                    case 4: return [2];
                }
            });
        });
    };
    LazyIterator.prototype.handleErrors = function (handler) {
        return new ErrorHandlingLazyIterator(this, handler);
    };
    LazyIterator.prototype.filter = function (predicate) {
        return new FilterIterator(this, predicate);
    };
    LazyIterator.prototype.map = function (transform) {
        return new MapIterator(this, transform);
    };
    LazyIterator.prototype.mapAsync = function (transform) {
        return new AsyncMapIterator(this, transform);
    };
    LazyIterator.prototype.serialMapAsync = function (transform) {
        return new AsyncMapIterator(this, transform).serial();
    };
    LazyIterator.prototype.flatmap = function (transform) {
        return new FlatmapIterator(this, transform);
    };
    LazyIterator.prototype.forEach = function (f) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.map(f).resolveFully()];
            });
        });
    };
    LazyIterator.prototype.serialForEach = function (f) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.serialMapAsync(f).resolveWhile(function (x) { return (x === true); })];
            });
        });
    };
    LazyIterator.prototype.rowMajorBatch = function (batchSize, smallLastBatch) {
        if (smallLastBatch === void 0) { smallLastBatch = true; }
        return new RowMajorBatchIterator(this, batchSize, smallLastBatch);
    };
    LazyIterator.prototype.columnMajorBatch = function (batchSize, smallLastBatch, zipFn) {
        if (smallLastBatch === void 0) { smallLastBatch = true; }
        if (zipFn === void 0) { zipFn = deep_map_1.zipToList; }
        var rowBatches = this.rowMajorBatch(batchSize, smallLastBatch);
        return rowBatches.map(function (x) { return deep_map_1.deepZip(x, zipFn); });
    };
    LazyIterator.prototype.concatenate = function (iterator, baseErrorHandler) {
        return new ChainedIterator(iteratorFromItems([this, iterator]), baseErrorHandler);
    };
    LazyIterator.prototype.take = function (count) {
        if (count < 0 || count == null) {
            return this;
        }
        return new TakeIterator(this, count);
    };
    LazyIterator.prototype.skip = function (count) {
        if (count < 0 || count == null) {
            return this;
        }
        return new SkipIterator(this, count);
    };
    LazyIterator.prototype.prefetch = function (bufferSize) {
        return new PrefetchIterator(this, bufferSize);
    };
    LazyIterator.prototype.shuffle = function (windowSize, seed) {
        return new ShuffleIterator(this, windowSize, seed);
    };
    LazyIterator.prototype.serial = function () {
        return new SerialIterator(this);
    };
    return LazyIterator;
}());
exports.LazyIterator = LazyIterator;
var ArrayIterator = (function (_super) {
    __extends(ArrayIterator, _super);
    function ArrayIterator(items) {
        var _this = _super.call(this) || this;
        _this.items = items;
        _this.trav = 0;
        return _this;
    }
    ArrayIterator.prototype.summary = function () {
        return "Array of " + this.items.length + " items";
    };
    ArrayIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var item, result;
            return __generator(this, function (_a) {
                if (this.trav >= this.items.length) {
                    return [2, { value: null, done: true }];
                }
                item = this.items[this.trav];
                if (item instanceof tf.Tensor) {
                    result = tf.clone(item);
                }
                else {
                    result = item;
                }
                this.trav++;
                return [2, { value: result, done: false }];
            });
        });
    };
    return ArrayIterator;
}(LazyIterator));
var FunctionCallIterator = (function (_super) {
    __extends(FunctionCallIterator, _super);
    function FunctionCallIterator(nextFn) {
        var _this = _super.call(this) || this;
        _this.nextFn = nextFn;
        return _this;
    }
    FunctionCallIterator.prototype.summary = function () {
        return "Function call";
    };
    FunctionCallIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                try {
                    return [2, this.nextFn()];
                }
                catch (e) {
                    e.message =
                        "Error thrown while iterating through a dataset: " + e.message;
                    throw e;
                }
                return [2];
            });
        });
    };
    return FunctionCallIterator;
}(LazyIterator));
var SerialIterator = (function (_super) {
    __extends(SerialIterator, _super);
    function SerialIterator(upstream) {
        var _this = _super.call(this) || this;
        _this.upstream = upstream;
        _this.lastRead = Promise.resolve({ value: null, done: false });
        return _this;
    }
    SerialIterator.prototype.summary = function () {
        return this.upstream.summary() + " -> Serial";
    };
    SerialIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.lastRead = this.lastRead.then(function () { return _this.serialNext(); });
                return [2, this.lastRead];
            });
        });
    };
    SerialIterator.prototype.serialNext = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2, this.upstream.next()];
            });
        });
    };
    return SerialIterator;
}(LazyIterator));
var SkipIterator = (function (_super) {
    __extends(SkipIterator, _super);
    function SkipIterator(upstream, maxCount) {
        var _this = _super.call(this) || this;
        _this.upstream = upstream;
        _this.maxCount = maxCount;
        _this.count = 0;
        _this.lastRead = Promise.resolve({ value: null, done: false });
        return _this;
    }
    SkipIterator.prototype.summary = function () {
        return this.upstream.summary() + " -> Skip";
    };
    SkipIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.lastRead = this.lastRead.then(function () { return _this.serialNext(); });
                return [2, this.lastRead];
            });
        });
    };
    SkipIterator.prototype.serialNext = function () {
        return __awaiter(this, void 0, void 0, function () {
            var skipped;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(this.count++ < this.maxCount)) return [3, 2];
                        return [4, this.upstream.next()];
                    case 1:
                        skipped = _a.sent();
                        if (skipped.done) {
                            return [2, skipped];
                        }
                        tf.dispose(skipped.value);
                        return [3, 0];
                    case 2: return [2, this.upstream.next()];
                }
            });
        });
    };
    return SkipIterator;
}(LazyIterator));
var TakeIterator = (function (_super) {
    __extends(TakeIterator, _super);
    function TakeIterator(upstream, maxCount) {
        var _this = _super.call(this) || this;
        _this.upstream = upstream;
        _this.maxCount = maxCount;
        _this.count = 0;
        return _this;
    }
    TakeIterator.prototype.summary = function () {
        return this.upstream.summary() + " -> Take";
    };
    TakeIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (this.count++ >= this.maxCount) {
                    return [2, { value: null, done: true }];
                }
                return [2, this.upstream.next()];
            });
        });
    };
    return TakeIterator;
}(LazyIterator));
var RowMajorBatchIterator = (function (_super) {
    __extends(RowMajorBatchIterator, _super);
    function RowMajorBatchIterator(upstream, batchSize, enableSmallLastBatch) {
        if (enableSmallLastBatch === void 0) { enableSmallLastBatch = true; }
        var _this = _super.call(this) || this;
        _this.upstream = upstream;
        _this.batchSize = batchSize;
        _this.enableSmallLastBatch = enableSmallLastBatch;
        _this.lastRead = Promise.resolve({ value: null, done: false });
        return _this;
    }
    RowMajorBatchIterator.prototype.summary = function () {
        return this.upstream.summary() + " -> RowMajorBatch";
    };
    RowMajorBatchIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.lastRead = this.lastRead.then(function () { return _this.serialNext(); });
                return [2, this.lastRead];
            });
        });
    };
    RowMajorBatchIterator.prototype.serialNext = function () {
        return __awaiter(this, void 0, void 0, function () {
            var batch, item;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        batch = [];
                        _a.label = 1;
                    case 1:
                        if (!(batch.length < this.batchSize)) return [3, 3];
                        return [4, this.upstream.next()];
                    case 2:
                        item = _a.sent();
                        if (item.done) {
                            if (this.enableSmallLastBatch && batch.length > 0) {
                                return [2, { value: batch, done: false }];
                            }
                            return [2, { value: null, done: true }];
                        }
                        batch.push(item.value);
                        return [3, 1];
                    case 3: return [2, { value: batch, done: false }];
                }
            });
        });
    };
    return RowMajorBatchIterator;
}(LazyIterator));
var FilterIterator = (function (_super) {
    __extends(FilterIterator, _super);
    function FilterIterator(upstream, predicate) {
        var _this = _super.call(this) || this;
        _this.upstream = upstream;
        _this.predicate = predicate;
        _this.lastRead = Promise.resolve({ value: null, done: false });
        return _this;
    }
    FilterIterator.prototype.summary = function () {
        return this.upstream.summary() + " -> Filter";
    };
    FilterIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.lastRead = this.lastRead.then(function () { return _this.serialNext(); });
                return [2, this.lastRead];
            });
        });
    };
    FilterIterator.prototype.serialNext = function () {
        return __awaiter(this, void 0, void 0, function () {
            var item;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!true) return [3, 2];
                        return [4, this.upstream.next()];
                    case 1:
                        item = _a.sent();
                        if (item.done || this.predicate(item.value)) {
                            return [2, item];
                        }
                        tf.dispose(item.value);
                        return [3, 0];
                    case 2: return [2];
                }
            });
        });
    };
    return FilterIterator;
}(LazyIterator));
var MapIterator = (function (_super) {
    __extends(MapIterator, _super);
    function MapIterator(upstream, transform) {
        var _this = _super.call(this) || this;
        _this.upstream = upstream;
        _this.transform = transform;
        return _this;
    }
    MapIterator.prototype.summary = function () {
        return this.upstream.summary() + " -> Map";
    };
    MapIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var item, inputTensors, mapped, outputTensors, _i, inputTensors_1, t;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.upstream.next()];
                    case 1:
                        item = _a.sent();
                        if (item.done) {
                            return [2, { value: null, done: true }];
                        }
                        inputTensors = tensor_util_1.getTensorsInContainer(item.value);
                        mapped = this.transform(item.value);
                        outputTensors = tensor_util_1.getTensorsInContainer(mapped);
                        for (_i = 0, inputTensors_1 = inputTensors; _i < inputTensors_1.length; _i++) {
                            t = inputTensors_1[_i];
                            if (!tensor_util_1.isTensorInList(t, outputTensors)) {
                                t.dispose();
                            }
                        }
                        return [2, { value: mapped, done: false }];
                }
            });
        });
    };
    return MapIterator;
}(LazyIterator));
var ErrorHandlingLazyIterator = (function (_super) {
    __extends(ErrorHandlingLazyIterator, _super);
    function ErrorHandlingLazyIterator(upstream, handler) {
        var _this = _super.call(this) || this;
        _this.upstream = upstream;
        _this.handler = handler;
        _this.count = 0;
        _this.lastRead = Promise.resolve({ value: null, done: false });
        return _this;
    }
    ErrorHandlingLazyIterator.prototype.summary = function () {
        return this.upstream.summary() + " -> handleErrors";
    };
    ErrorHandlingLazyIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.lastRead = this.lastRead.then(function () { return _this.serialNext(); });
                return [2, this.lastRead];
            });
        });
    };
    ErrorHandlingLazyIterator.prototype.serialNext = function () {
        return __awaiter(this, void 0, void 0, function () {
            var e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!true) return [3, 5];
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4, this.upstream.next()];
                    case 2: return [2, _a.sent()];
                    case 3:
                        e_1 = _a.sent();
                        if (!this.handler(e_1)) {
                            return [2, { value: null, done: true }];
                        }
                        return [3, 4];
                    case 4: return [3, 0];
                    case 5: return [2];
                }
            });
        });
    };
    return ErrorHandlingLazyIterator;
}(LazyIterator));
var AsyncMapIterator = (function (_super) {
    __extends(AsyncMapIterator, _super);
    function AsyncMapIterator(upstream, transform) {
        var _this = _super.call(this) || this;
        _this.upstream = upstream;
        _this.transform = transform;
        return _this;
    }
    AsyncMapIterator.prototype.summary = function () {
        return this.upstream.summary() + " -> AsyncMap";
    };
    AsyncMapIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var item, inputTensors, mapped, outputTensors, _i, inputTensors_2, t;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.upstream.next()];
                    case 1:
                        item = _a.sent();
                        if (item.done) {
                            return [2, { value: null, done: true }];
                        }
                        inputTensors = tensor_util_1.getTensorsInContainer(item.value);
                        return [4, this.transform(item.value)];
                    case 2:
                        mapped = _a.sent();
                        outputTensors = tensor_util_1.getTensorsInContainer(mapped);
                        for (_i = 0, inputTensors_2 = inputTensors; _i < inputTensors_2.length; _i++) {
                            t = inputTensors_2[_i];
                            if (!tensor_util_1.isTensorInList(t, outputTensors)) {
                                t.dispose();
                            }
                        }
                        return [2, { value: mapped, done: false }];
                }
            });
        });
    };
    return AsyncMapIterator;
}(LazyIterator));
var OneToManyIterator = (function (_super) {
    __extends(OneToManyIterator, _super);
    function OneToManyIterator() {
        var _this = _super.call(this) || this;
        _this.outputQueue = new growing_ring_buffer_1.GrowingRingBuffer();
        _this.lastRead = Promise.resolve({ value: null, done: false });
        return _this;
    }
    OneToManyIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.lastRead = this.lastRead.then(function () { return _this.serialNext(); });
                return [2, this.lastRead];
            });
        });
    };
    OneToManyIterator.prototype.serialNext = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(this.outputQueue.length() === 0)) return [3, 2];
                        return [4, this.pump()];
                    case 1:
                        if (!(_a.sent())) {
                            return [2, { value: null, done: true }];
                        }
                        return [3, 0];
                    case 2: return [2, { value: this.outputQueue.shift(), done: false }];
                }
            });
        });
    };
    return OneToManyIterator;
}(LazyIterator));
exports.OneToManyIterator = OneToManyIterator;
var FlatmapIterator = (function (_super) {
    __extends(FlatmapIterator, _super);
    function FlatmapIterator(upstream, transform) {
        var _this = _super.call(this) || this;
        _this.upstream = upstream;
        _this.transform = transform;
        return _this;
    }
    FlatmapIterator.prototype.summary = function () {
        return this.upstream.summary() + " -> Flatmap";
    };
    FlatmapIterator.prototype.pump = function () {
        return __awaiter(this, void 0, void 0, function () {
            var item, inputTensors, mappedArray, outputTensors, _i, inputTensors_3, t;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.upstream.next()];
                    case 1:
                        item = _a.sent();
                        if (item.done) {
                            return [2, false];
                        }
                        inputTensors = tensor_util_1.getTensorsInContainer(item.value);
                        mappedArray = this.transform(item.value);
                        outputTensors = tensor_util_1.getTensorsInContainer(mappedArray);
                        this.outputQueue.pushAll(mappedArray);
                        for (_i = 0, inputTensors_3 = inputTensors; _i < inputTensors_3.length; _i++) {
                            t = inputTensors_3[_i];
                            if (!tensor_util_1.isTensorInList(t, outputTensors)) {
                                t.dispose();
                            }
                        }
                        return [2, true];
                }
            });
        });
    };
    return FlatmapIterator;
}(OneToManyIterator));
var ChainedIterator = (function (_super) {
    __extends(ChainedIterator, _super);
    function ChainedIterator(iterators, baseErrorHandler) {
        var _this = _super.call(this) || this;
        _this.baseErrorHandler = baseErrorHandler;
        _this.lastRead = null;
        _this.iterator = null;
        _this.moreIterators = iterators;
        return _this;
    }
    ChainedIterator.prototype.summary = function () {
        var upstreamSummaries = 'TODO: fill in upstream of chained summaries';
        return upstreamSummaries + " -> Chained";
    };
    ChainedIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.lastRead = this.readFromChain(this.lastRead);
                return [2, this.lastRead];
            });
        });
    };
    ChainedIterator.prototype.readFromChain = function (lastRead) {
        return __awaiter(this, void 0, void 0, function () {
            var iteratorResult, itemResult;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, lastRead];
                    case 1:
                        _a.sent();
                        if (!(this.iterator == null)) return [3, 3];
                        return [4, this.moreIterators.next()];
                    case 2:
                        iteratorResult = _a.sent();
                        if (iteratorResult.done) {
                            return [2, { value: null, done: true }];
                        }
                        this.iterator = iteratorResult.value;
                        if (this.baseErrorHandler != null) {
                            this.iterator = this.iterator.handleErrors(this.baseErrorHandler);
                        }
                        _a.label = 3;
                    case 3: return [4, this.iterator.next()];
                    case 4:
                        itemResult = _a.sent();
                        if (itemResult.done) {
                            this.iterator = null;
                            return [2, this.readFromChain(lastRead)];
                        }
                        return [2, itemResult];
                }
            });
        });
    };
    return ChainedIterator;
}(LazyIterator));
exports.ChainedIterator = ChainedIterator;
var ZipMismatchMode;
(function (ZipMismatchMode) {
    ZipMismatchMode[ZipMismatchMode["FAIL"] = 0] = "FAIL";
    ZipMismatchMode[ZipMismatchMode["SHORTEST"] = 1] = "SHORTEST";
    ZipMismatchMode[ZipMismatchMode["LONGEST"] = 2] = "LONGEST";
})(ZipMismatchMode = exports.ZipMismatchMode || (exports.ZipMismatchMode = {}));
var ZipIterator = (function (_super) {
    __extends(ZipIterator, _super);
    function ZipIterator(iterators, mismatchMode) {
        if (mismatchMode === void 0) { mismatchMode = ZipMismatchMode.FAIL; }
        var _this = _super.call(this) || this;
        _this.iterators = iterators;
        _this.mismatchMode = mismatchMode;
        _this.count = 0;
        _this.currentPromise = null;
        return _this;
    }
    ZipIterator.prototype.summary = function () {
        var upstreamSummaries = 'TODO: fill in upstream of zip summaries';
        return "{" + upstreamSummaries + "} -> Zip";
    };
    ZipIterator.prototype.nextState = function (afterState) {
        return __awaiter(this, void 0, void 0, function () {
            function getNext(container) {
                if (container instanceof LazyIterator) {
                    var result = container.next();
                    return {
                        value: result.then(function (x) {
                            numIterators++;
                            if (x.done) {
                                iteratorsDone++;
                            }
                            return x.value;
                        }),
                        recurse: false
                    };
                }
                else {
                    return { value: null, recurse: true };
                }
            }
            var numIterators, iteratorsDone, mapped;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, afterState];
                    case 1:
                        _a.sent();
                        numIterators = 0;
                        iteratorsDone = 0;
                        return [4, deep_map_1.deepMapAndAwaitAll(this.iterators, getNext)];
                    case 2:
                        mapped = _a.sent();
                        if (numIterators === iteratorsDone) {
                            return [2, { value: null, done: true }];
                        }
                        if (iteratorsDone > 0) {
                            switch (this.mismatchMode) {
                                case ZipMismatchMode.FAIL:
                                    throw new Error('Zipped streams should have the same length. ' +
                                        ("Mismatched at element " + this.count + "."));
                                case ZipMismatchMode.SHORTEST:
                                    return [2, { value: null, done: true }];
                                case ZipMismatchMode.LONGEST:
                                default:
                            }
                        }
                        this.count++;
                        return [2, { value: mapped, done: false }];
                }
            });
        });
    };
    ZipIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.currentPromise = this.nextState(this.currentPromise);
                        return [4, this.currentPromise];
                    case 1: return [2, (_a.sent())];
                }
            });
        });
    };
    return ZipIterator;
}(LazyIterator));
var PrefetchIterator = (function (_super) {
    __extends(PrefetchIterator, _super);
    function PrefetchIterator(upstream, bufferSize) {
        var _this = _super.call(this) || this;
        _this.upstream = upstream;
        _this.bufferSize = bufferSize;
        _this.buffer = new ring_buffer_1.RingBuffer(bufferSize);
        return _this;
    }
    PrefetchIterator.prototype.summary = function () {
        return this.upstream.summary() + " -> Prefetch";
    };
    PrefetchIterator.prototype.refill = function () {
        while (!this.buffer.isFull()) {
            var v = this.upstream.next();
            this.buffer.push(v);
        }
    };
    PrefetchIterator.prototype.next = function () {
        this.refill();
        return this.buffer.shift();
    };
    return PrefetchIterator;
}(LazyIterator));
exports.PrefetchIterator = PrefetchIterator;
var ShuffleIterator = (function (_super) {
    __extends(ShuffleIterator, _super);
    function ShuffleIterator(upstream, windowSize, seed) {
        var _this = _super.call(this, upstream, windowSize) || this;
        _this.upstream = upstream;
        _this.windowSize = windowSize;
        _this.upstreamExhausted = false;
        _this.random = seedrandom.alea(seed || tf.util.now().toString());
        _this.lastRead = Promise.resolve({ value: null, done: false });
        return _this;
    }
    ShuffleIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.lastRead = this.lastRead.then(function () { return _this.serialNext(); });
                return [2, this.lastRead];
            });
        });
    };
    ShuffleIterator.prototype.randomInt = function (max) {
        return Math.floor(this.random() * max);
    };
    ShuffleIterator.prototype.chooseIndex = function () {
        return this.randomInt(this.buffer.length());
    };
    ShuffleIterator.prototype.serialNext = function () {
        return __awaiter(this, void 0, void 0, function () {
            var chosenIndex, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.upstreamExhausted) {
                            this.refill();
                        }
                        _a.label = 1;
                    case 1:
                        if (!!this.buffer.isEmpty()) return [3, 3];
                        chosenIndex = this.chooseIndex();
                        return [4, this.buffer.shuffleExcise(chosenIndex)];
                    case 2:
                        result = _a.sent();
                        if (result.done) {
                            this.upstreamExhausted = true;
                        }
                        else {
                            this.refill();
                            return [2, result];
                        }
                        return [3, 1];
                    case 3: return [2, { value: null, done: true }];
                }
            });
        });
    };
    return ShuffleIterator;
}(PrefetchIterator));
exports.ShuffleIterator = ShuffleIterator;
